l1-contracts/src/core/Decoder.sol

```diff
@@ -89,41 +110,33 @@ contract Decoder {
       uint256 l2BlockNumber,
       bytes32 startStateHash,
       bytes32 endStateHash,
-      bytes32 publicInputHash
+      bytes32 publicInputHash,
+      bytes32[] memory l2ToL1Msgs,
+      bytes32[] memory l1ToL2Msgs
     )
   {
     l2BlockNumber = _getL2BlockNumber(_l2Block);
     // Note, for startStateHash to match the storage, the l2 block number must be new - 1.
     // Only jumping 1 block at a time.
     startStateHash = _computeStateHash(l2BlockNumber - 1, 0x4, _l2Block);
     endStateHash = _computeStateHash(l2BlockNumber, 0x120, _l2Block);
-    publicInputHash = _computePublicInputsHash(_l2Block);
-  }
 
-  /**
-   * Computes a hash of the public inputs from the calldata
-   * @param _l2Block - The L2 block calldata.
-   * @return sha256(header[0x4: 0x23c], diffRoot, l1Tol2MessagesHash)
-   */
-  function _computePublicInputsHash(bytes calldata _l2Block) internal pure returns (bytes32) {
-    // header size - block number size + one value for the diffRoot + one value for l1ToL2MessagesHash
-    uint256 size = 0x23c - 0x04 + 0x20 + 0x20;
+    // Compute consumables used by public input hash or returned to rollup
+    bytes32 diffRoot;
+    bytes32 l1ToL2messagesHash;
+    (diffRoot, l1ToL2messagesHash, l2ToL1Msgs, l1ToL2Msgs) = _computeConsumables(_l2Block);
 
     // Compute the public inputs hash
+    // header size - block number size + one value for the diffRoot + one value for l1ToL2MessagesHash
+    uint256 size = 0x23c - 0x04 + 0x20 + 0x20;
     bytes memory temp = new bytes(size);
     assembly {
       calldatacopy(add(temp, 0x20), add(_l2Block.offset, 0x04), size)
-    }
-
-    // Diff root
-    (bytes32 diffRoot, bytes32 l1ToL2messagesHash) = _computeDiffRootAndMessagesHash(_l2Block);
-    assembly {
       let endOfTreesData := sub(0x23c, 0x04)
       mstore(add(temp, add(0x20, endOfTreesData)), diffRoot)
       mstore(add(temp, add(0x40, endOfTreesData)), l1ToL2messagesHash)
     }
-
-    return bytes32(uint256(sha256(temp)) % P);
+    publicInputHash = bytes32(uint256(sha256(temp)) % P);
```

Comment:
sorry to be annoying again but this asm can be in its own function with the diffRoot and msgHash passed in